File: agents/backend/onyx/server/features/robot_movement_ai/aws/modules/plugins/plugin_manager.py
"""
Plugin Manager
==============

Advanced plugin management system.
"""

import logging
import importlib
from typing import Dict, List, Any, Optional, Type
from dataclasses import dataclass
from abc import ABC, abstractmethod

logger = logging.getLogger(__name__)


@dataclass
class PluginMetadata:
    """Plugin metadata."""
    name: str
    version: str
    description: str
    author: str
    dependencies: List[str] = None
    enabled: bool = True


class IPlugin(ABC):
    """Plugin interface."""
    
    @abstractmethod
    def initialize(self, config: Dict[str, Any]) -> bool:
        """Initialize plugin."""
        pass
    
    @abstractmethod
    def get_metadata(self) -> PluginMetadata:
        """Get plugin metadata."""
        pass
    
    @abstractmethod
    def execute(self, *args, **kwargs) -> Any:
        """Execute plugin."""
        pass
    
    @abstractmethod
    def cleanup(self):
        """Cleanup plugin."""
        pass


class PluginManager:
    """Plugin manager."""
    
    def __init__(self):
        self._plugins: Dict[str, IPlugin] = {}
        self._metadata: Dict[str, PluginMetadata] = {}
        self._configs: Dict[str, Dict[str, Any]] = {}
    
    def register(
        self,
        plugin: IPlugin,
        config: Optional[Dict[str, Any]] = None
    ):
        """Register plugin."""
        metadata = plugin.get_metadata()
        
        if metadata.name in self._plugins:
            logger.warning(f"Plugin {metadata.name} already registered, replacing...")
        
        self._plugins[metadata.name] = plugin
        self._metadata[metadata.name] = metadata
        self._configs[metadata.name] = config or {}
        
        if metadata.enabled:
            plugin.initialize(self._configs[metadata.name])
        
        logger.info(f"Registered plugin: {metadata.name} v{metadata.version}")
    
    def load_from_module(self, module_path: str, plugin_class: str):
        """Load plugin from module."""
        try:
            module = importlib.import_module(module_path)
            plugin_class_obj = getattr(module, plugin_class)
            plugin = plugin_class_obj()
            self.register(plugin)
        except Exception as e:
            logger.error(f"Failed to load plugin from {module_path}: {e}")
    
    def enable(self, plugin_name: str):
        """Enable plugin."""
        if plugin_name in self._plugins:
            self._plugins[plugin_name].initialize(self._configs[plugin_name])
            self._metadata[plugin_name].enabled = True
            logger.info(f"Enabled plugin: {plugin_name}")
    
    def disable(self, plugin_name: str):
        """Disable plugin."""
        if plugin_name in self._plugins:
            self._plugins[plugin_name].cleanup()
            self._metadata[plugin_name].enabled = False
            logger.info(f"Disabled plugin: {plugin_name}")
    
    def get_plugin(self, plugin_name: str) -> Optional[IPlugin]:
        """Get plugin by name."""
        return self._plugins.get(plugin_name)
    
    def list_plugins(self) -> List[PluginMetadata]:
        """List all plugins."""
        return list(self._metadata.values())
    
    def execute_plugin(self, plugin_name: str, *args, **kwargs) -> Any:
        """Execute plugin."""
        plugin = self.get_plugin(plugin_name)
        if not plugin:
            raise ValueError(f"Plugin {plugin_name} not found")
        
        if not self._metadata[plugin_name].enabled:
            raise ValueError(f"Plugin {plugin_name} is disabled")
        
        return plugin.execute(*args, **kwargs)
    
    def cleanup_all(self):
        """Cleanup all plugins."""
        for plugin in self._plugins.values():
            try:
                plugin.cleanup()
            except Exception as e:
                logger.error(f"Error cleaning up plugin: {e}")

File: character_clothing_changer_ai/static/js/core/plugin-manager.js
/**
 * Plugin Manager Module
 * =====================
 * Manages plugins and extensions for the application
 */

const PluginManager = {
    /**
     * Registered plugins
     */
    plugins: new Map(),
    
    /**
     * Plugin hooks
     */
    hooks: new Map(),
    
    /**
     * Register a plugin
     */
    register(name, plugin) {
        if (!plugin || typeof plugin !== 'object') {
            throw new Error(`Invalid plugin: ${name}`);
        }
        
        if (!plugin.name) {
            plugin.name = name;
        }
        
        if (!plugin.version) {
            plugin.version = '1.0.0';
        }
        
        // Validate plugin
        this.validatePlugin(plugin);
        
        // Initialize plugin if init method exists
        if (typeof plugin.init === 'function') {
            try {
                plugin.init();
            } catch (error) {
                if (typeof Logger !== 'undefined') {
                    Logger.error(`Error initializing plugin ${name}:`, error);
                }
                throw error;
            }
        }
        
        this.plugins.set(name, plugin);
        
        // Register plugin hooks
        if (plugin.hooks) {
            Object.keys(plugin.hooks).forEach(hookName => {
                this.registerHook(hookName, plugin.hooks[hookName], name);
            });
        }
        
        if (typeof Logger !== 'undefined') {
            Logger.info(`Plugin registered: ${name} v${plugin.version}`);
        }
        
        // Emit plugin registered event
        if (typeof EventBus !== 'undefined') {
            EventBus.emit('plugin:registered', { name, plugin });
        }
        
        return this;
    },
    
    /**
     * Unregister a plugin
     */
    unregister(name) {
        const plugin = this.plugins.get(name);
        if (!plugin) {
            return false;
        }
        
        // Cleanup plugin if cleanup method exists
        if (typeof plugin.cleanup === 'function') {
            try {
                plugin.cleanup();
            } catch (error) {
                if (typeof Logger !== 'undefined') {
                    Logger.error(`Error cleaning up plugin ${name}:`, error);
                }
            }
        }
        
        // Remove plugin hooks
        this.hooks.forEach((handlers, hookName) => {
            this.hooks.set(
                hookName,
                handlers.filter(h => h.plugin !== name)
            );
        });
        
        this.plugins.delete(name);
        
        if (typeof Logger !== 'undefined') {
            Logger.info(`Plugin unregistered: ${name}`);
        }
        
        // Emit plugin unregistered event
        if (typeof EventBus !== 'undefined') {
            EventBus.emit('plugin:unregistered', { name });
        }
        
        return true;
    },
    
    /**
     * Get a plugin
     */
    get(name) {
        return this.plugins.get(name);
    },
    
    /**
     * Check if plugin is registered
     */
    has(name) {
        return this.plugins.has(name);
    },
    
    /**
     * Get all plugins
     */
    getAll() {
        return Array.from(this.plugins.values());
    },
    
    /**
     * Register a hook
     */
    registerHook(hookName, handler, pluginName) {
        if (!this.hooks.has(hookName)) {
            this.hooks.set(hookName, []);
        }
        
        this.hooks.get(hookName).push({
            handler,
            plugin: pluginName
        });
        
        if (typeof Logger !== 'undefined') {
            Logger.debug(`Hook registered: ${hookName} by ${pluginName}`);
        }
    },
    
    /**
     * Execute a hook
     */
    executeHook(hookName, ...args) {
        if (!this.hooks.has(hookName)) {
            return args[0]; // Return first argument if no hooks
        }
        
        const handlers = this.hooks.get(hookName);
        let result = args[0];
        
        handlers.forEach(({ handler, plugin }) => {
            try {
                const hookResult = handler(result, ...args.slice(1));
                if (hookResult !== undefined) {
                    result = hookResult;
                }
            } catch (error) {
                if (typeof Logger !== 'undefined') {
                    Logger.error(`Error executing hook ${hookName} in plugin ${plugin}:`, error);
                }
            }
        });
        
        return result;
    },
    
    /**
     * Execute hook asynchronously
     */
    async executeHookAsync(hookName, ...args) {
        if (!this.hooks.has(hookName)) {
            return args[0];
        }
        
        const handlers = this.hooks.get(hookName);
        let result = args[0];
        
        for (const { handler, plugin } of handlers) {
            try {
                const hookResult = await handler(result, ...args.slice(1));
                if (hookResult !== undefined) {
                    result = hookResult;
                }
            } catch (error) {
                if (typeof Logger !== 'undefined') {
                    Logger.error(`Error executing async hook ${hookName} in plugin ${plugin}:`, error);
                }
            }
        }
        
        return result;
    },
    
    /**
     * Validate plugin structure
     */
    validatePlugin(plugin) {
        if (!plugin.name) {
            throw new Error('Plugin must have a name');
        }
        
        if (!plugin.version) {
            throw new Error('Plugin must have a version');
        }
        
        // Optional: validate plugin API version compatibility
        if (plugin.apiVersion) {
            const currentAPIVersion = '1.0.0'; // Current API version
            // Add version compatibility check if needed
        }
        
        return true;
    },
    
    /**
     * Get plugin information
     */
    getPluginInfo(name) {
        const plugin = this.plugins.get(name);
        if (!plugin) {
            return null;
        }
        
        return {
            name: plugin.name,
            version: plugin.version,
            description: plugin.description,
            author: plugin.author,
            enabled: plugin.enabled !== false,
            hooks: Array.from(this.hooks.entries())
                .filter(([_, handlers]) => handlers.some(h => h.plugin === name))
                .map(([hookName]) => hookName)
        };
    },
    
    /**
     * Enable/disable plugin
     */
    setPluginEnabled(name, enabled) {
        const plugin = this.plugins.get(name);
        if (!plugin) {
            return false;
        }
        
        plugin.enabled = enabled;
        
        if (typeof Logger !== 'undefined') {
            Logger.info(`Plugin ${name} ${enabled ? 'enabled' : 'disabled'}`);
        }
        
        // Emit plugin state change event
        if (typeof EventBus !== 'undefined') {
            EventBus.emit('plugin:state_changed', { name, enabled });
        }
        
        return true;
    },
    
    /**
     * Get all hooks
     */
    getHooks() {
        return Array.from(this.hooks.keys());
    },
    
    /**
     * Clear all plugins
     */
    clear() {
        const pluginNames = Array.from(this.plugins.keys());
        pluginNames.forEach(name => this.unregister(name));
        this.hooks.clear();
        
        if (typeof Logger !== 'undefined') {
            Logger.info('All plugins cleared');
        }
    }
};

// Export for use in other modules
if (typeof module !== 'undefined' && module.exports) {
    module.exports = PluginManager;
}

